File: archive/src/main/java/com/dipanjal/leetcode/linkedlist/merge_two_sorted_list.py
# https://leetcode.com/problems/merge-two-sorted-lists

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    
    def mergeTwoLists(self, list1, list2):
        dummy_head = ListNode(-1)
        tail = dummy_head
        
        while list1 and list2:
            if list1.val < list2.val:
                # add to tail
                tail.next = list1
                # move forward 
                list1 = list1.next
            else:
                # add to tail
                tail.next = list2
                # move forward 
                list2 = list2.next
            
            tail = tail.next
        
        # collect the remaining and append with the tail of the output
        if list1:
            tail.next = list1
        if list2:
            tail.next = list2

        return dummy_head.next

File: archive/dipanjal/leetcode/linkedlist/MergeTwoSortedList.java
package com.dipanjal.leetcode.linkedlist;

/**
 * @author dipanjal
 * @since 0.0.1
 * https://leetcode.com/problems/merge-two-sorted-lists/
 * easy
 */
public class MergeTwoSortedList {
    public static class ListNode {
       int val;
       ListNode next;
       ListNode() {}
       ListNode(int val) { this.val = val; }
       ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode newHead = new ListNode(-1);
        ListNode tail = newHead;

        while(l1 != null && l2 != null) {
            if(l1.val < l2.val){
                tail.next = l1;
                l1 = l1.next;
            }else {
                tail.next = l2;
                l2 = l2.next;
            }
            tail = tail.next;
        }

        /** Now collect the reminders if has */
        /* means: L1 Has some nodes remains to be collected */
        if(l1 != null)
            tail.next = l1;

        /* means: L2 Has some nodes remains to be collected */
        if(l2 != null)
            tail.next = l2;

        return newHead.next; //because we want to ignore the initial dummyHead -1
    }

    public static void travarse(ListNode head) {
        while(head != null) {
            String arrow = head.next != null ? "->" : "";
            System.out.print(head.val+arrow);
            head = head.next;
        }
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
        ListNode l2 = new ListNode(1, new ListNode(3, new ListNode(4)));

        ListNode merged = mergeTwoLists(l1, l2);
        travarse(merged);
    }
}
